### PR TITLE
RDM-2252 Altered Get Roles implementation

### PR DESCRIPTION
**Pull request by tzarsmango**
Altered Get Roles implementation to handle multiple requests in a single request
Fixed some minor sonar issues

https://tools.hmcts.net/jira/browse/RDM-2252

This task is second phase of RDM-1923. Will prevent tons of requests and exceptions thrown when a role cannot be found.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java [hunk 1]
```diff
@@ -91,6 +91,10 @@ public String userRoleClassification() {
         return caseDefinitionHost + "/api/user-role?role={userRole}";
     }
 
+    public String userRolesClassificationsURL() {
+        return caseDefinitionHost + "/api/user-roles/{roles}";
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java]: I would like to open a new story to rename this endpoint in definition store. Which I find confusing. 
This really should be 
`/api/security-classifications/{roles}`

What do you think Fatih?

File: src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java [hunk 2]
```diff
@@ -49,6 +47,18 @@ public UserRole getUserRoleClassifications(String userRole) {
         return userRoleClassifications.computeIfAbsent(userRole, caseDefinitionRepository::getUserRoleClassifications);
     }
 
+    @Override
+    public List<UserRole> getClassificationsForUserRoleList(List<String> userRoles) {
+        final List<String> missingRoles = userRoles.stream()
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java]: can we rename missingRoles to missingClassifications?

File: src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java [hunk 3]
```diff
@@ -49,6 +47,18 @@ public UserRole getUserRoleClassifications(String userRole) {
...
+            .filter(role -> !userRoleClassifications.containsKey(role))
+            .collect(Collectors.toList());
+        if (!missingRoles.isEmpty()) {
+            final List<UserRole> missingClassifications = caseDefinitionRepository.getClassificationsForUserRoleList(missingRoles);
+            missingClassifications.forEach(userRole -> userRoleClassifications.putIfAbsent(userRole.getRole(), userRole));
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java]: can we rename userRole to classification?

File: src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java [hunk 4]
```diff
@@ -112,28 +107,43 @@ public UserRole getUserRoleClassifications(String userRole) {
...
             }
         }
     }
 
+    @Override
+    public List<UserRole> getClassificationsForUserRoleList(List<String> userRoles) {
+        try {
+            final HttpEntity requestEntity = new HttpEntity<CaseType>(securityUtils.authorizationHeaders());
+            final Map<String, String> queryParams = new HashMap<>();
+            Optional<String> roles = userRoles.stream().reduce((s1, s2) -> s1 + "," + s2);
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java]: perhaps StringUtils.join(slist, ',') is simpler?

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java [hunk 5]
```diff
@@ -76,10 +75,11 @@ public IDAMProperties getUserDetails() {
     @Override
     public Set<SecurityClassification> getUserClassifications(String jurisdictionId) {
         Set<String> roles = this.getUserRoles();
+        final List<String> filteredRoles = roles.stream()
```
Comment (mario-paniccia) [inline src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: I've seen this filtering of roles on many other places.
Can we introduce a method getJurisdictionRoles(String jurisdiction) and move this code in there?

File: src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java (hunk 1 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java]: https://tools.hmcts.net/jira/browse/RDM-2319

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java (hunk 5 above)
Comment (mario-paniccia) [reply src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: Also, we can further improve performance by excluding roles of type -loa* . Can we introduce a method 
getJurisdictionRolesExcludingLoa and use that?

File: src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java (hunk 1 above)
Comment (tzarsmango) [reply src/main/java/uk/gov/hmcts/ccd/ApplicationParams.java]: that looks better, thanks

File: src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java (hunk 2 above)
Comment (tzarsmango) [reply src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java]: Actually that name is used a few lines below that one

File: src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java (hunk 4 above)
Comment (tzarsmango) [reply src/main/java/uk/gov/hmcts/ccd/data/definition/DefaultCaseDefinitionRepository.java]: yes 👍

File: src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java (hunk 3 above)
Comment (tzarsmango) [reply src/main/java/uk/gov/hmcts/ccd/data/definition/CachedCaseDefinitionRepository.java]: done

File: src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java (hunk 5 above)
Comment (tzarsmango) [reply src/main/java/uk/gov/hmcts/ccd/data/user/DefaultUserRepository.java]: Actually there's already a filterRoles private method used a little below that line, but I think it's filtering roles with `caseworker-JURISTDICTION` pattern and not filtering out the `-loa` roles...

**Pull request merged**